### ISSUE TITLE
Why Trax client is the socket server?

### ISSUE THREAD

**Issue by cybertk** [votchallenge/trax#30]
In current trax implementation, **trax_client** is socket server, listening on specified TCP port, while **trax_server** connecting to the **trax_client**.

From the meaning of C/S, **trax_server** should be the one listening on a socket, why the truth is not?

**Comment by cybertk** [comment]
Actually, we are designing a end-to-end system based on **Trax** protocal. We want our trackers wait their clients after start up.

**Comment by lukacu** [comment]
The terminology client/server came from the way the information is processed: the client sends requests, the server does some work. The actual formation of a communication channel is not a part of the protocol itself and another approach could be implemented. I agree that the current approach is inconvenient for some scenarios, but it was implemented to support classical evaluation. If you have another idea it can be added to the library.

**Comment by cybertk** [comment]
Cool, that make sense. Trax already supports socket at communication level. What do you suggest if we add another kind of socket communication?

**Comment by lukacu** [comment]
Well, if you would like to be able to run tracker first, then you have two options. In both you have to extend trax_server_setup function to either support server socket (this would be probably messy since all the code for this is currently in the client module) or to keep trying to connect to a given port for some time.

But more importantly the client would not have control over tracker process in this case which has to be handled in the client class (so no process reset, no explicit termination). I think that this should be done and is a great idea but at the moment I do not have to work on it so contributions of any kind or discussion is welcome.

**Comment by cybertk** [comment]
> you have to extend trax_server_setup function to either support server socket (this would be probably messy since all the code for this is currently in the client module)

You are right, it will be very messy if extending existing `trax_server_setup`. So I think we should bring a new separate logic to support server socket, parallel with existing two method, fd and socket. Perhaps call it **server_socket**?

> But more importantly the client would not have control over tracker process in this case which has to be handled in the client class (so no process reset, no explicit termination).

I thinks we can still control tracker process in this case, i.e. for process reset, tracker(actually in libtrax part) can kill itself and re-spawn after received `quit` signal.

**Comment by lukacu** [comment]
> I thinks we can still control tracker process in this case, i.e. for process reset, tracker(actually in libtrax part) can kill itself and re-spawn after received quit signal.

Well if a tracker behaves ok, then this is not a problem. It is a problem when a tracker freezes or something like this, then you usually have to kill it manually.

**Comment by cybertk** [comment]
The purpose is using Trax on the non-evaluation domain, I'm not sure if it will break the design purpose of Trax.

**Comment by lukacu** [comment]
No, it is just an usability issue, but is not related to the protocol itself. If you will write your own client infrastructure then I think that the proposed solution is very likely fine.
